README.md
# plugin-AldirBlanc #
Plugin que implementa no Mapas Culturais os formulários do inciso I e II da Lei Aldir Blanc 

## Configuração do fluxo de homologação e validações das inscrições
Para o fluxo de homologação e validações, são utilizados 3 plugins:

- [plugin-AldirBlanc](https://github.com/mapasculturais/plugin-AldirBlanc) - base da solução Aldir Blanc, implementa o cadastro e a base dos plugins validadores
- [plugin-AldirBlancDataprev](https://github.com/mapasculturais/plugin-AldirBlancDataprev) - implementa as exportações dos CSVs para o Dataprev e as importações dos arquivos de retorno. 
- [plugin-AldirBlancValidador](https://github.com/mapasculturais/plugin-AldirBlancValidador) - validador genérico no qual é possível configurar o layout do arquivo exportado. O arquivo exportado contém duas colunas para serem preenchidas com o status e as observações das avaliações, para que seja enviado para o sistema novamente. Este plugin pode ser instanciado várias vezes para diferentes validações.

É possível configurar no sistema a ordem em que ocorrem as validações e a homologação das inscrições, além de quando que as avaliações/validações são consolidadas nas inscrições para que possam ser publicadas para o requerente. Por padrão, primeiro ocorre a homologação e depois as validações.

Abaixo seguem 2 exemplos de configuração, com os parâmetros comentados. 

### Exemplo 1 
Neste exemplo de configuração, primeiro ocorre a homologação e depois ocorrem em paralelo as validações do Dataprev e SCGE (plugin-AldirBlancValidador). A consolidação do resultado só acontece quando Dataprev e SCGE tiverem validado a inscrição.

- homologação
    - validação Dataprev
    - validação SCGE (Validador Genérico)

```PHP

    'plugins' => [
        
        'AldirBlanc' => [
            'namespace' => 'AldirBlanc',
            'config' => [ 
                ... outras configurações 

                // primeiro ocorre a homologação, só depois as validações
                'homologacao_requer_validacao' => [],

                // indica que a consolidação do status "Selecionada" só deve acontecer 
                // depois das validações
                'consolidacao_requer_validacao' => ['dataprev', 'scge']
            ]
        ],

        'AldirBlancDataprev' => [
            'namespace' => 'AldirBlancDataprev',
            'config' => [
                // indica que só deve exportar as inscrições já homologadas
                'exportador_requer_homologacao' => true,

                // indica que só deve consolidar o resultado se a inscrição
                // já tiver sido processada pelo SCGE
                'consolidacao_requer_validacoes' => ['scge']
            ],
        ],

        'SCGE' => [
            'namespace' => 'AldirBlancValidador',
            'config' => [
                // slug utilizado como id do controller e identificador do validador
                'slug' => 'scge',

                // nome apresentado na interface
                'name' => 'SCGE',

                // indica que só deve exportar as inscrições já homologadas
                'exportador_requer_homologacao' => true,

                // indica que a exportação não requer nenhuma validação
                'exportador_requer_validacao' => [],

                // indica que só deve consolidar o resultado se a inscrição
                // já tiver sido processada pelo Dataprev
                'consolidacao_requer_validacoes' => ['dataprev'],

                'inciso1' => [
                    // id do field do formulário 
                    'CPF' => 33,

                    // para campos compostos, ou que não estão no formulário, 
                    // é possível utilizar uma função que recebe a inscrição 
                    // e o nome da coluna, no caso $key = 'campo_personalizado'
                    'campo_personalizado' => function ($registration, $key) {
                        return $registration->owner->
                    }
                ],
            ]
        ]
```


### Exemplo 2 
Neste exemplo de configuração, primeiro ocorre a validação pelo Dataprev, depois a homologação e só depois a validação do SCGE (plugin-AldirBlancValidador). A consolidação do resultado só acontece quando o SCGE validar a inscrição. Somente as inscrições validadas pelo dataprev aparecerão para os avaliadores avaliarem.

- validação Dataprev
    - homologação
        - validação SCGE (Validador Genérico)

```PHP

    'plugins' => [
        
        'AldirBlanc' => [
            'namespace' => 'AldirBlanc',
            'config' => [ 
                ... outras configurações 

                // primeiro ocorre a validação pelo Dataprev, só depois a homologação
                'homologacao_requer_validacao' => ['dataprev'],

                // indica que a consolidação do status "Selecionada" só deve acontecer 
                // depois das validações
                'consolidacao_requer_validacao' => ['dataprev', 'scge']
            ]
        ],

        'AldirBlancDataprev' => [
            'namespace' => 'AldirBlancDataprev',
            'config' => [
                // indica que deve exportar TODAS as inscrições enviadas (status pendente)
                'exportador_requer_homologacao' => false,

                // indica que só deve consolidar o resultado se a inscrição
                // já tiver sido processada pelo SCGE
                'consolidacao_requer_validacoes' => ['scge']
            ],
        ],

        'SCGE' => [
            'namespace' => 'AldirBlancValidador',
            'config' => [
                // slug utilizado como id do controller e identificador do validador
                'slug' => 'scge',

                // nome apresentado na interface
                'name' => 'SCGE',

                // indica que só deve exportar as inscrições já homologadas
                'exportador_requer_homologacao' => true,

                // indica que a exportação requer que as inscrições tenham sido validadas
                // pelo Dataprev
                'exportador_requer_validacao' => ['dataprev'],

                // indica que deve consolidar o resultado na importação do arquivo
                // de retorno.
                'consolidacao_requer_validacoes' => [],

                'inciso1' => [
                    // id do field do formulário 
                    'CPF' => 33,

                    // para campos compostos, ou que não estão no formulário, 
                    // é possível utilizar uma função que recebe a inscrição 
                    // e o nome da coluna, no caso $key = 'campo_personalizado'
                    'campo_personalizado' => function ($registration, $key) {
                        return $registration->owner->
                    }
                ],
            ]
        ]
```


CHANGELOG.md
## [Unreleased]
- Endpoint para corrigir espaços e agentes de inscrições mediadas
- Interface para envio dos e-mails dos PPG10x
- Implementa exportador do Banpará [#185]

## [v2.4.0]
- Corrige bug no processamento de múltiplos pagamentos do exportador PPG100 [#169](https://github.com/mapasculturais/plugin-AldirBlanc/issues/169)
- Corrige exibição de listagem de arquivos que apareciam em oportunidades que não são da AldirBlanc
- Recuperação de senha de mediados
- Corrige envio de e-mails dos PPG10x e adiciona suporte a duas mensagens para o caso geral
- Corrige endereço de e-mail exibido na mensagem de status dos PPG10x
- Adiciona reply-to aos e-mails dos PPG10x e altera a fonte do endereço de e-mail destino
- Atualiza envio de e-mails e exibição da mensagem de status para adequar ao novo fluxo dos PPG10x
- Endpoint para validação dos dados Bancários [#174](https://github.com/mapasculturais/plugin-AldirBlanc/issues/174)
- Altera referência padrão dos PPG10x para ID do pagamento [#162](https://github.com/mapasculturais/plugin-AldirBlanc/issues/162)
- Corrige bug de armazenamento dos metadados no importador do MCI470 [#171](https://github.com/mapasculturais/plugin-AldirBlanc/issues/171)
- Realiza correções no importador do MCI470 [#162](https://github.com/mapasculturais/plugin-AldirBlanc/issues/162)
- Padroniza exibição da área de upload de desbancarizados [#162](https://github.com/mapasculturais/plugin-AldirBlanc/issues/162)
- Adiciona suporte a múltiplos pagamentos no mesmo arquivo PPG100 [#169](https://github.com/mapasculturais/plugin-AldirBlanc/issues/169)
- Adiciona ao formulário de avaliação dos incisos I e II a informação da consolidação atual das avaliações da inscrição [#163](https://github.com/mapasculturais/plugin-AldirBlanc/issues/163)
- Realiza correções no importador do MCI470 [#162](https://github.com/mapasculturais/plugin-AldirBlanc/issues/162)
- Refatorar CNAB240 para exportar valores quebrados EX.: 2.904,80 (Ref. [#167](https://github.com/mapasculturais/plugin-AldirBlanc/issues/167))
- Refatorar calculo da data de pagamento para 5 dias após o envio (Ref. [#167](https://github.com/mapasculturais/plugin-AldirBlanc/issues/167))
- Inserir Número da inscrição no logradouro para que a mesma retorne nos arquivos de retorno BB (Ref. [#167](https://github.com/mapasculturais/plugin-AldirBlanc/issues/167))
- Corrige código de Câmara centralizadora para outros banco para 018 no CNAB240 (Ref. [#170](https://github.com/mapasculturais/plugin-AldirBlanc/issues/170))
- Corrige Forma de lançamento para lote de outros bancos para 041 no CNAB240(Ref. [#170](https://github.com/mapasculturais/plugin-AldirBlanc/issues/170))
- Corrige prefixo para casos de conta corrente de 510 para 51 no CNAB240(Ref. [#170](https://github.com/mapasculturais/plugin-AldirBlanc/issues/170))
- Inserire opção de fazer exportação de arquivos teste (TS) no CNAB240(Ref. [#170](https://github.com/mapasculturais/plugin-AldirBlanc/issues/170))
- Inserire configurações para habilitar e desabilitar botão do CNAB240 nos incisos no CNAB240(Ref. [#170](https://github.com/mapasculturais/plugin-AldirBlanc/issues/170))
- Inserire posições 178 a 179 campo 24.3A No lote 2 Poupança BB inserir o valor default 11 no CNAB240(Ref. [#172](https://github.com/mapasculturais/plugin-AldirBlanc/issues/172))
- Corrige complemento de conta poupança para 9 caracteres prefixo 51 + 7 números1 no CNAB240(Ref. [#172](https://github.com/mapasculturais/plugin-AldirBlanc/issues/172))

- Refatorar importador dataprev para inscrições que foram em um primeiro momento inseridas na plataforma como Não processadas, sejam reprocessadas caso em um segundo momento a mesma venha com status de processado (Ref. [#19](https://github.com/mapasculturais/plugin-AldirBlanc/issues/19))
- Refatorar código para que o busque cnab240 os dados bancários nos metadados inseridos na tabela agent_meta apos a validação de contas bancárias. (Ref. [#175](https://github.com/mapasculturais/plugin-AldirBlanc/issues/175))
- Insere quebra de linha no treiller do lote no CNAB240 (Ref. [#177](https://github.com/mapasculturais/plugin-AldirBlanc/issues/177)
- Refatorar código para que o busque cnab240 os dados bancários nos metadados inseridos na tabela agent_meta apos a validação de contas bancárias. (Ref. [#175](https://github.com/mapasculturais/plugin-AldirBlanc/issues/175))
- Refatora Importador CNAB240 para que busque as inscrições pelo CPF. (Ref. [#185](https://github.com/mapasculturais/plugin-AldirBlanc/issues/185))
- Refatora Importador CNAB240 para permitir baixar arquivos de consolidação. (Ref. [#185](https://github.com/mapasculturais/plugin-AldirBlanc/issues/185))
- Refatora Importador CNAB240 para qpermitir processar mais de um retorno para mesma inscrição, fazendo a consolidação do resultado. (Ref. [#185](https://github.com/mapasculturais/plugin-AldirBlanc/issues/185))
- Refatora CNAB240 para buscar DV de Agencia e Contas em fields separados. (Ref. [#182](https://github.com/mapasculturais/plugin-AldirBlanc/issues/182))

## [v2.3.1] - 2020-12-01
- Implementa os importadores para o MCI470 e PPG101 [#162](https://github.com/mapasculturais/plugin-AldirBlanc/issues/162)
- Reordena as avaliações antes da reconsolidação do resultado das inscrições, para colocar as avaliações que foram importadas (as que têm id igual ao id da inscrição) para serem processadas primeiro;
- Implementa função para processar os retornos de pagamento do Banco do Brasil para o CNAB240 [#173]https://github.com/mapasculturais/plugin-AldirBlanc/issues/173 
- Refatora CNAB240 para ser possivel separar a exportação por lotes [#176]https://github.com/mapasculturais/plugin-AldirBlanc/issues/176 
- Insere quebra de linha no treiller do lote no CNAB240 (Ref. [#177](https://github.com/mapasculturais/plugin-AldirBlanc/issues/177)

config-ppg10x.php
<?php

/**
 * Configuração do exportador do PPG100
 *
 * ppg100Serial: temporariamente usado para trocar o número seqüencial do
 *               arquivo PPG100
 * fieldMap: usado para localizar os diversos campos na oportunidade
 * header: configuração do header; lista seqüencial de campos com length, type e
 *         name, mais default se for um campo constante ou function para um
 *         campo dinâmico
 * details: configuração dos detalhes; este formato tem apenas um tipo de
 *          detalhe mas na configuração é lista por compatibilidade
 * detail: fields com lista seqüencial de campos com length, type, e name, mais
 *         default se for um campo constante; se o tipo do campo for "meta", o
 *         campo deve conter a lista de subcampos fields e uma function para
 *         preenchimento
 * trailer: configuração do trailer; lista seqüencial de campos com length, type
 *          e name, mais default se for um campo constante; os nomes dos campos
 *          dinâmicos são utilizados como chaves do último parâmetro passado
 *          para o método que gera o trailer
 * condition: estrutura de verificações para decidir se o registro entra na
 *            remessa
 *
 * Tudo o que for "function" precisa ser o nome de um método no controller.
 * ToDo: documentar assinaturas para esses métodos.
 */
return [
    "serial" => 0,
    "idMap" => "CSV/ppgIdMap.csv", // comentar a linha para mapear pelo id do pagamento
    "fieldMap" => [
        "wantsPaymentOrder" => "field_8564",
        "singleParent" => "field_119",
        //"numeroProtocolo" => "id",
        "senhaSaque" => "number",
        "cpf" => "field_104",
        "indicadorAcao" => "id",
    ],
    "header" => [
        [
            "length" => 1,
            "type" => "int",
            "name" => "tipoRegistro",
            "default" => 0,
        ],
        [
            "length" => 8,
            "type" => "int",
            "name" => "dataRemessa",
            "function" => "genericDateDDMMYYYY",
        ],
        [
            "length" => 4,
            "type" => "int",
            "name" => "horaRemessa",
            "function" => "genericTimeHHMM",
        ],
        [
            "length" => 6,
            "type" => "text",
            "name" => "nomeArquivo",
            "default" => "PPG100",
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "mciCliente",
            "default" => 0, // preenchimento SECULT
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "codigoParametroCliente",
            "default" => 0, // preenchimento SECULT
        ],
        [
            "length" => 3,
            "type" => "int",
            "name" => "versaoLayout",
            "default" => 11,
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "sequencialRemessa",
            "function" => "sequenceNumber",
        ],
        [
            "length" => 9,
            "type" => "text",
            "name" => "padding",
            "default" => "",
        ],
        [
            "length" => 4,
            "type" => "int",
            "name" => "agenciaDebito",
            "default" => 0, // preenchimento SECULT
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "contaDebito",
            "default" => 0, // preenchimento SECULT
        ],
        [
            "length" => 120,
            "type" => "text",
            "name" => "padding",
            "default" => "",
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "numeroRegistro",
            "default" => 1,
        ],
    ],
    "details" => [
        [
            "fields" => [
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "default" => 1,
                ],
                [
                    "length" => 15,
                    "type" => "meta",
                    "name" => "numeroProtocolo", // enviar CSV no extraData
                    "fields" => [
                        [ // apenas no novo formato
                            "length" => 6,
                            "type" => "int",
                            "name" => "padding",
                            "default" => 0,
                        ],
                        [
                            "length" => 3, // 3 no novo formato, 4 no antigo
                            "type" => "int",
                            "name" => "idBB",
                            "default" => 0, // preenchimento SECULT
                        ],
                        [
                            "length" => 6, // 6 no novo formato, 10 no antigo
                            "type" => "int",
                            "name" => "idCliente",
                        ],
                        // [ // apenas no antigo formato
                        //     "length" => 1,
                        //     "type" => "int",
                        //     "name" => "dv",
                        // ],
                    ],
                    //"function" => "ppg100ProtocolNumberPA",
                ],
                [
                    "length" => 6,
                    "type" => "int",
                    "name" => "senhaSaque",
                    "function" => "ppg100PIN",
                ],
                [
                    "length" => 10,
                    "type" => "text",
                    "name" => "padding",
                    "default" => "",
                ],
                [
                    "length" => 11,
                    "type" => "int",
                    "name" => "cpf",
                ],
                [
                    "length" => 11,
                    "type" => "int",
                    "name" => "valorCarga",
                ],
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "indicadorAcao",
                    "default" => 4, // sempre abertura e carga pois cada pagamento é um protocolo
                ],
                [
                    "length" => 136,
                    "type" => "text",
                    "name" => "padding",
                    "default" => "",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "numeroRegistro",
                ],
            ],
        ],
    ],
    "trailer" => [
        [
            "length" => 1,
            "type" => "int",
            "name" => "tipoRegistro",
            "default" => 9,
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "totalDetalhes",
        ],
        [
            "length" => 11,
            "type" => "int", // em desacordo com a documentação
            "name" => "totalCarga",
        ],
        [
            "length" => 170,
            "type" => "text",
            "name" => "padding",
            "default" => "",
        ],
        [
            "length" => 9,
            "type" => "int",
            "name" => "numeroRegistro",
        ],
    ],
    "condition" => [
        "operator" => "and",
        "operands" => [
            [
                "operator" => "exists",
                "operands" => ["wantsPaymentOrder"],
            ],
            [
                "operator" => "prefix",
                "operands" => [
                    "wantsPaymentOrder",
                    ["const" => "Ordem de pagamento"],
                ],
            ],
            // [
            //     "operator" => "equals",
            //     "operands" => ["singleParent", ["const" => "NÃO"]]
            // ],
        ],
    ],
    "return" => [
        "topLevel" => [
            [
                "length" => 1,
                "type" => "int",
                "name" => "tipoRegistro",
                "map" => "records",
            ],
            [
                "length" => 190,
                "type" => "text",
                "name" => "data",
            ],
            [
                "length" => 9,
                "type" => "int",
                "name" => "numeroRegistro",
            ],
        ],
        "records" => [
            "0" => [ // header
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 0,
                ],
                [
                    "length" => 8,
                    "type" => "int",
                    "name" => "dataRemessa",
                ],
                [
                    "length" => 4,
                    "type" => "int",
                    "name" => "horaRemessa",
                ],
                [
                    "length" => 6,
                    "type" => "text",
                    "name" => "nomeArquivo",
                    "match" => "PPG101",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "mciCliente",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "codigoParametroCliente",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 3,
                    "type" => "int",
                    "name" => "versaoLayout",
                    "match" => 11,
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "sequencialRemessa",
                    "field" => "sequenceNumber",
                ],
                [
                    "length" => 4,
                    "type" => "int",
                    "name" => "agenciaDebito",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "contaDebito",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 5,
                    "type" => "int",
                    "name" => "codigoRecusa",
                    "capture" => "fileResultCode",
                ],
                [
                    "length" => 40,
                    "type" => "text",
                    "name" => "textoRecusa",
                    "capture" => "fileResultMessage",
                ],
                [
                    "length" => 84,
                    "type" => "text",
                    "name" => "padding",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "numeroRegistro",
                    "match" => 1,
                ],
            ],
            "1" => [ // detail
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 1,
                ],
                [ // apenas no novo formato
                    "length" => 6,
                    "type" => "int",
                    "name" => "padding",
                    "default" => 0,
                ],
                [
                    "length" => 3, // 3 no novo formato, 4 no antigo
                    "type" => "int",
                    "name" => "idBB",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 6, // 6 no novo formato, 10 no antigo
                    "type" => "int",
                    "name" => "idCliente",
                    "capture" => "reference",
                ],
                // [ // apenas no antigo formato
                //     "length" => 1,
                //     "type" => "int",
                //     "name" => "dv",
                // ],
                [
                    "length" => 5,
                    "type" => "int",
                    "name" => "codigoRecusa",
                    "capture" => "paymentCode",
                ],
                [
                    "length" => 40,
                    "type" => "text",
                    "name" => "textoRecusa",
                    "capture" => "paymentMessage",
                ],
                [
                    "length" => 130,
                    "type" => "text",
                    "name" => "padding",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "numeroRegistro",
                ],
            ],
            "9" => [ // trailer
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 9,
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "totalAceitos",
                    "capture" => "countAccepted",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "totalRejeitados",
                    "capture" => "countRejected",
                ],
                [
                    "length" => 172,
                    "type" => "text",
                    "name" => "padding",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "numeroRegistro",
                ],
            ],
        ],
    ],
    "followup" => [
        "topLevel" => [
            [
                "length" => 1,
                "type" => "int",
                "name" => "tipoRegistro",
                "map" => "records",
            ],
            [
                "length" => 69,
                "type" => "text",
                "name" => "data",
            ],
        ],
        "records" => [
            "0" => [ // header
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 0,
                ],
                [
                    "length" => 8,
                    "type" => "int",
                    "name" => "dataRemessa",
                ],
                [
                    "length" => 6,
                    "type" => "int",
                    "name" => "horaRemessa",
                ],
                [
                    "length" => 6,
                    "type" => "text",
                    "name" => "nomeArquivo",
                    "match" => "PPG102",
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "mciCliente",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 4,
                    "type" => "int",
                    "name" => "numeroContrato",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 36,
                    "type" => "text",
                    "name" => "padding",
                ],
            ],
            "1" => [ // detail
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 1,
                ],
                [ // apenas no novo formato
                    "length" => 6,
                    "type" => "int",
                    "name" => "padding",
                    "match" => 0,
                ],
                [
                    "length" => 3, // 3 no novo formato, 4 no antigo
                    "type" => "int",
                    "name" => "idBB",
                    "match" => 0, // preenchimento SECULT
                ],
                [
                    "length" => 6, // 6 no novo formato, 10 no antigo
                    "type" => "int",
                    "name" => "idCliente",
                    "capture" => "reference",
                ],
                // [ // apenas no antigo formato
                //     "length" => 1,
                //     "type" => "int",
                //     "name" => "dv",
                // ],
                [
                    "length" => 11,
                    "type" => "int",
                    "name" => "cpf",
                    "capture" => "cpf",
                ],
                [
                    "length" => 17,
                    "type" => "int",
                    "name" => "valorMovimento",
                    "capture" => "operationAmount",
                ],
                [
                    "length" => 1,
                    "type" => "text",
                    "name" => "tipoMovimento",
                    "capture" => "operationType",
                ],
                [
                    "length" => 17,
                    "type" => "int",
                    "name" => "saldoDisponivel",
                    "capture" => "currentBalance",
                ],
                [
                    "length" => 1,
                    "type" => "text",
                    "name" => "situacaoSenha",
                    "capture" => "statusPIN",
                ],
                [
                    "length" => 7,
                    "type" => "text",
                    "name" => "padding",
                ],
            ],
            "9" => [ // trailer
                [
                    "length" => 1,
                    "type" => "int",
                    "name" => "tipoRegistro",
                    "match" => 9,
                ],
                [
                    "length" => 9,
                    "type" => "int",
                    "name" => "totalRegistros",
                    "capture" => "countEntries",
                ],
                [
                    "length" => 60,
                    "type" => "text",
                    "name" => "padding",
                ],
            ],
        ],
    ],
];
